src/NuGet.Clients/NuGet.VisualStudio.OnlineEnvironment.Client/RestoreCommandHandler.cs

```diff
@@ -0,0 +1,106 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.ServiceHub.Framework;
+using Microsoft.VisualStudio;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Shell.Interop;
+using Microsoft.VisualStudio.Shell.ServiceBroker;
+using Microsoft.VisualStudio.Threading;
+using NuGet.VisualStudio.Internal.Contracts;
+using IAsyncServiceProvider = Microsoft.VisualStudio.Shell.IAsyncServiceProvider;
+using Task = System.Threading.Tasks.Task;
+
+namespace NuGet.VisualStudio.OnlineEnvironment.Client
+{
+    internal class RestoreCommandHandler
+    {
+        private readonly JoinableTaskFactory _joinableTaskFactory;
+        private readonly IAsyncServiceProvider _asyncServiceProvider;
+        private CancellationTokenSource _cancelBuildToken;
+
+        public RestoreCommandHandler(JoinableTaskFactory joinableTaskFactory, IAsyncServiceProvider asyncServiceProvider)
+        {
+            _joinableTaskFactory = joinableTaskFactory ?? throw new ArgumentNullException(nameof(joinableTaskFactory));
+            _asyncServiceProvider = asyncServiceProvider ?? throw new ArgumentNullException(nameof(asyncServiceProvider));
+        }
+
+        public void RunSolutionRestore()
+        {
+            _joinableTaskFactory.RunAsync(async () =>
+            {
```

Comment:
I'd like it better if there was a method `RunSoltuionRestoreAsync`, and have `RunSoltuionRestore` do nothing more than `_joinableTaskFactory.RunAsync(RunSolutionRestoreAsync)`.

Could we even consider moving the `JTF.RunAsync` to the calling location, to make it clearer that this class is not "legacy being migrated to async", but is designed to be async, and the calling location is more clearly synchronous and needs to handle calling async methods?

Reply:
I can do that.